catalog/app/utils/json-schema/json-schema.ts

```diff
@@ -0,0 +1,221 @@
+    format: string
+    items: JsonSchema
+    properties: Record<string, JsonSchema>
+    type: string | string[] | JsonSchema[]
+  } & Record<CompoundCondition, JsonSchema[]>
+>
+
+// NOTE: can't use R.prop directly, because R.prop type doesn't allow undefined
+const prop = (
+  property: keyof JsonSchema,
+  optSchema?: JsonSchema,
+): JsonSchema[keyof JsonSchema] => (optSchema ? R.prop(property, optSchema) : undefined)
+
+export const isSchemaArray = (optSchema?: JsonSchema) =>
+  prop('type', optSchema) === 'array'
+
+export const isSchemaObject = (optSchema?: JsonSchema) =>
+  prop('type', optSchema) === 'object'
+
+const isSchemaString = (optSchema?: JsonSchema) => prop('type', optSchema) === 'string'
+
+const isSchemaNumber = (optSchema?: JsonSchema) => prop('type', optSchema) === 'number'
+
+const isSchemaInteger = (optSchema?: JsonSchema) => prop('type', optSchema) === 'integer'
+
+export const isSchemaBoolean = (optSchema?: JsonSchema) =>
+  prop('type', optSchema) === 'boolean'
+
+const isSchemaNull = (optSchema?: JsonSchema) => prop('type', optSchema) === 'null'
+
+export const isSchemaEnum = (optSchema?: JsonSchema) => !!prop('enum', optSchema)
+
+export const isSchemaOneOf = (optSchema?: JsonSchema) => !!prop('oneOf', optSchema)
+
+export const isSchemaAnyOf = (optSchema?: JsonSchema) => !!prop('anyOf', optSchema)
+
+export const isSchemaAllOf = (optSchema?: JsonSchema) => !!prop('allOf', optSchema)
+
+const isSchemaConst = (optSchema?: JsonSchema) => !!prop('const', optSchema)
+
+function isSchemaCompound(optSchema?: JsonSchema) {
+  if (!optSchema) return false
+  return ['anyOf', 'oneOf', 'not', 'allOf'].some(
+    (key) => optSchema[key as 'anyOf' | 'oneOf' | 'not' | 'allOf'],
+  )
+}
+
+const isSchemaReference = (optSchema: JsonSchema) => !!prop('$ref', optSchema)
+
+export const isNestedType = R.either(isSchemaArray, isSchemaObject)
+
+function compoundTypeToHumanString(
+  optSchema: JsonSchema,
+  condition: CompoundCondition,
+  divider: string,
+): string {
+  if (!Array.isArray(prop(condition, optSchema))) return ''
+
+  return (optSchema[condition] as JsonSchema[])
+    .map(schemaTypeToHumanString)
+    .filter((v) => v !== 'undefined') // NOTE: sic, see default case of `schemaTypeToHumanString`
+    .join(divider)
+}
+
+export function schemaTypeToHumanString(optSchema?: JsonSchema) {
+  if (!optSchema) return ''
+  return R.cond<JsonSchema, string>([
+    [isSchemaEnum, () => 'enum'],
+    [isSchemaConst, () => 'const'],
+    [isSchemaBoolean, () => 'boolean'],
+    [isSchemaNull, () => 'null'],
+    // NOTE: enum and const can be string too,
+    //       that's why they are first
+    [
+      R.propOr('', 'type'),
+      () =>
+        Array.isArray(optSchema.type)
+          ? optSchema.type.join('|')
+          : (optSchema.type as string),
+    ],
+    [isSchemaAnyOf, () => compoundTypeToHumanString(optSchema, 'anyOf', '|')],
+    [isSchemaOneOf, () => compoundTypeToHumanString(optSchema, 'oneOf', '&')],
```

Comment:
what's the difference btw `anyOf` and `oneOf` in this context? seems like both these cases could use `|` and leave `&` for `allOf`